src/main/java/appeng/me/storage/BasicCellInventory.java

```diff
@@ -23,50 +6,90 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.nbt.NBTTagCompound;
 
-import appeng.api.AEApi;
 import appeng.api.config.Actionable;
 import appeng.api.exceptions.AppEngException;
 import appeng.api.implementations.items.IStorageCell;
 import appeng.api.networking.security.IActionSource;
-import appeng.api.storage.ICellInventoryHandler;
-import appeng.api.storage.IMEInventory;
+import appeng.api.storage.ICellInventory;
 import appeng.api.storage.ISaveProvider;
 import appeng.api.storage.IStorageChannel;
-import appeng.api.storage.channels.IItemStorageChannel;
 import appeng.api.storage.data.IAEItemStack;
+import appeng.api.storage.data.IAEStack;
 import appeng.core.AEConfig;
 import appeng.core.AELog;
+import appeng.util.item.AEStack;
 
 
-public class ItemCellInventory extends AbstractCellInventory<IAEItemStack>
+public class BasicCellInventory<T extends IAEStack<T>> extends AbstractCellInventory<T>
 {
-	private ItemCellInventory( final ItemStack o, final ISaveProvider container ) throws AppEngException
+	private final IStorageChannel<T> channel;
+
+	private BasicCellInventory( final IStorageCell<T> cellType, final ItemStack o, final ISaveProvider container )
 	{
-		super( o, container, 8 );
+		super( cellType, o, container );
+		this.channel = cellType.getChannel();
 	}
 
-	public static ICellInventoryHandler<IAEItemStack> getCell( final ItemStack o, final ISaveProvider container2 )
+	public static <T extends IAEStack<T>> ICellInventory<T> createInventory( final ItemStack o, final ISaveProvider container )
 	{
 		try
 		{
-			return new ItemCellInventoryHandler( new ItemCellInventory( o, container2 ) );
+			if( o == null )
+			{
+				throw new AppEngException( "ItemStack was used as a cell, but was not a cell!" );
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "ItemStack was used as a cell, but was not a cell!" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1192)